src/codegen/dispatch.rs

```diff
@@ -8,59 +8,74 @@ use crate::codegen::{
     Builtin, Expression, Options,
 };
 use crate::{
-    sema::ast::{Namespace, Parameter, Type},
+    sema::ast::{Namespace, StructType, Type},
     Target,
 };
 use num_bigint::{BigInt, Sign};
 use num_traits::Zero;
-use solang_parser::pt;
-use solang_parser::pt::{FunctionTy, Loc};
-use std::sync::Arc;
+use solang_parser::{pt, pt::Loc};
 
 /// Create the dispatch for the Solana target
 pub(super) fn function_dispatch(
     contract_no: usize,
     all_cfg: &[ControlFlowGraph],
     ns: &mut Namespace,
+    opt: &Options,
 ) -> ControlFlowGraph {
     let mut vartab = Vartable::new(ns.next_id);
-    let mut cfg = ControlFlowGraph::new(
-        format!("dispatch_{}", ns.contracts[contract_no].name),
-        ASTFunction::None,
-    );
+    let mut cfg = ControlFlowGraph::new("solang_dispatch".into(), ASTFunction::None);
+
+    let switch_block = cfg.new_basic_block("switch".to_string());
+    let no_function_matched = cfg.new_basic_block("no_function_matched".to_string());
+
+    let argsdata_var = vartab.temp_name("input", &Type::BufferPointer);
+    let argslen_var = vartab.temp_name("input_len", &Type::Uint(64));
 
-    cfg.params = Arc::new(vec![
-        Parameter {
+    let sol_params =
+        Expression::FunctionArg(Loc::Codegen, Type::Struct(StructType::SolParameters), 0);
+
+    // argsdata_var = load ty:bufferptr (structmember ty:bufferptr (funcarg ty:struct(solparam), 2))
+    cfg.add(
+        &mut vartab,
+        Instr::Set {
+            res: argsdata_var,
             loc: Loc::Codegen,
-            id: None,
-            ty: Type::BufferPointer,
-            ty_loc: None,
-            indexed: false,
-            readonly: false,
-            recursive: false,
+            expr: Expression::Load(
+                Loc::Codegen,
+                Type::BufferPointer,
+                Expression::StructMember(
+                    Loc::Codegen,
+                    Type::BufferPointer,
+                    sol_params.clone().into(),
+                    2,
+                )
+                .into(),
+            ),
         },
-        Parameter {
+    );
+
+    let argsdata = Expression::Variable(Loc::Codegen, Type::BufferPointer, argsdata_var);
+
+    // argslen_var = load ty:uint64 (structmember ty:uin64 (funcarg ty:struct(solparam), 3))
+    cfg.add(
+        &mut vartab,
+        Instr::Set {
+            res: argslen_var,
             loc: Loc::Codegen,
-            id: None,
-            ty: Type::Uint(64),
-            ty_loc: None,
-            indexed: false,
-            readonly: false,
-            recursive: false,
+            expr: Expression::Load(
+                Loc::Codegen,
+                Type::Uint(64),
+                Expression::StructMember(Loc::Codegen, Type::Uint(64), sol_params.into(), 3).into(),
```

Comment:
Shouldn't the type here be a `Type::Ref(Type::Uint(64)`? If it is not a reference, there is no need to load it.